### PR TITLE
bpo-39156: Break up COMPARE_OP into four logically distinct opcodes.

### PR DESCRIPTION
**Pull request by markshannon**
Breaks up COMPARE_OP into four logically distinct opcodes:

*   COMPARE_OP for rich comparisons
*   IS_OP for 'is' and 'is not' tests
*   CONTAINS_OP for 'in' and 'is not' tests
*   JUMP_IF_NOT_EXC_MATCH for checking exceptions in 'try-except' statements.

This improves the clarity of the interpreter and should provide a modest speedup.

https://bugs.python.org/issue39156

### CODE REVIEW

File: Doc/library/dis.rst [hunk 1]
```diff
@@ -961,6 +975,13 @@ All of the following opcodes use their arguments.
 
    .. versionadded:: 3.1
 
+.. opcode:: JUMP_IF_NOT_EXC_MATCH (target)
+
+    Tests whether the second value on the stack is an exception matching TOS, and jumps if it is not.
```
Comment (ZackerySpytz) [inline Doc/library/dis.rst]: This should be wrapped at 80 chars.

File: Python/ceval.c [hunk 2]
```diff
@@ -2896,19 +2895,95 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
...
+
+#define CANNOT_CATCH_MSG "catching classes that do not inherit from "\
+                         "BaseException is not allowed"
+
+        case TARGET(JUMP_IF_NOT_EXC_MATCH): {
+            PyObject *right = POP();
+            PyObject *left = POP();
+            if (PyTuple_Check(right)) {
+                Py_ssize_t i, length;
+                length = PyTuple_Size(right);
```
Comment (ZackerySpytz) [inline Python/ceval.c]: I think `PyTuple_GET_SIZE()` should be used here.

File: Python/ceval.c [hunk 3]
```diff
@@ -2896,19 +2895,95 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
...
+                for (i = 0; i < length; i += 1) {
```
Comment (ZackerySpytz) [inline Python/ceval.c]: i++

File: Python/ceval.c [hunk 4]
```diff
@@ -2896,19 +2895,95 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
...
+                                    CANNOT_CATCH_MSG);
+                    Py_DECREF(left);
+                    Py_DECREF(right);
+                    goto error;
+                }
+            }
+            int res = PyErr_GivenExceptionMatches(left, right);
+            Py_DECREF(left);
+            Py_DECREF(right);
+            if (res > 0) {
```
Comment (ZackerySpytz) [inline Python/ceval.c:2975]: IMO, this check for `res > 0` should not be included.

**Review by ZackerySpytz — COMMENTED** [review]
I have some minor comments.

File: Python/ceval.c (hunk 4 above)
Comment (markshannon) [reply Python/ceval.c:2975]: Why?

**Pull request merged**